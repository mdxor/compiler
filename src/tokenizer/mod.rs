mod atx_heading;
mod indented_code;
mod rule;
mod setext_heading;
mod thematic_break;
mod token;
