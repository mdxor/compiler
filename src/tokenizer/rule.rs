pub const ATX_HEADING_RULE: &str = "^(#{1,6})(?:| (.*))(?:\n|$)";
pub const ATX_HEADING_CLOSING_RULE: &str = "(^| )#+ *$";
pub const THEMATIC_BREAK_RULE: &str = r"((?:- *){3,}|(?:_ *){3,}|(?:\* *){3,})(?:\n+|$)";
