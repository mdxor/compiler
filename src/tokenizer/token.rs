#[cfg(test)]
use serde::Serialize;

#[derive(Debug, PartialEq)]
#[cfg_attr(test, derive(Serialize))]
pub struct TokenResult<'source> {
  pub size: usize,
  pub token: Token<'source>,
}

#[derive(Debug, PartialEq)]
#[cfg_attr(test, derive(Serialize))]
pub enum Token<'source> {
  ATXHeading(ATXHeading<'source>),
  ThematicBreak,
  SetextHeading(SetextHeading<'source>),
  IndentedCode(IndentedCode<'source>),
  FencedCode(FencedCode<'source>),
}

#[derive(Debug, PartialEq)]
#[cfg_attr(test, derive(Serialize))]
pub struct ATXHeading<'source> {
  pub level: usize,
  pub raw_inlines: &'source str,
}

#[derive(Debug, PartialEq)]
#[cfg_attr(test, derive(Serialize))]
pub struct SetextHeading<'source> {
  pub level: usize,
  pub raw: &'source str,
}

#[derive(Debug, PartialEq)]
#[cfg_attr(test, derive(Serialize))]
pub struct IndentedCode<'source> {
  pub codes: Vec<&'source str>,
}

#[derive(Debug, PartialEq)]
#[cfg_attr(test, derive(Serialize))]
pub struct FencedCode<'source> {
  pub code: &'source str,
  pub meta_string: &'source str,
  pub language: &'source str,
}
