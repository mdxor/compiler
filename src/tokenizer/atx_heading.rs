use crate::tokenizer::rule;
use crate::tokenizer::token::*;
use lazy_static::lazy_static;
use regex::Regex;
pub fn atx_heading(source: &str) -> Option<TokenResult> {
  lazy_static! {
    static ref RULE: Regex = Regex::new(rule::ATX_HEADING_RULE).unwrap();
    static ref CLOSING_RULE: Regex = Regex::new(rule::ATX_HEADING_CLOSING_RULE).unwrap();
  }
  if let Some(caps) = RULE.captures(source) {
    let size = caps.get(0).unwrap().as_str().len();
    let level = caps.get(1).unwrap().as_str().len();
    let mut raw_inlines = caps.get(2).map_or("", |v| v.as_str());
    if let Some(caps) = CLOSING_RULE.captures(raw_inlines) {
      let closing_size = caps.get(0).map_or(0, |v| v.as_str().len());
      raw_inlines = &raw_inlines[..raw_inlines.len() - closing_size];
    }
    Some(TokenResult {
      size,
      token: Token::ATXHeading(ATXHeading { level, raw_inlines }),
    })
  } else {
    None
  }
}

#[test]
fn test_atx_heading() {
  let cases = vec![
    "# foo",
    "## foo",
    "### foo",
    "#### foo",
    "##### foo",
    "###### foo",
    "####### foo",
    "#5 bolt",
    "#hashtag",
    r"\## foo",
    r"# foo *bar* \*baz\*",
    "###",
    "# foo#",
    "# foo #",
    "# #",
  ];
  let mut results = vec![];
  for case in cases {
    results.push(atx_heading(case));
  }
  insta::assert_yaml_snapshot!(results);
}
