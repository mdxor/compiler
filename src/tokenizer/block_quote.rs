use crate::tokenizer::rule;
use crate::tokenizer::token::*;
use lazy_static::lazy_static;
use regex::Regex;
pub fn block_quote(source: &str) -> Option<TokenResult> {
  lazy_static! {
    static ref RULE: Regex = Regex::new(rule::BLOCK_QUOTE).unwrap();
  }
  if let Some(caps) = RULE.captures(source) {
    let size = caps.get(0).unwrap().as_str().len();
    Some(TokenResult {
      size,
      token: Token::BlockQuote,
    })
  } else {
    None
  }
}

#[test]
fn test_block_quote() {
  let cases = vec![">", "> "];
  let mut results = vec![];
  for case in cases {
    results.push(block_quote(case));
  }
  insta::assert_yaml_snapshot!(results);
}
