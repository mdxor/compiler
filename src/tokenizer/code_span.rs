use crate::tokenizer::rule;
use crate::tokenizer::token::*;
use lazy_static::lazy_static;
use regex::Regex;
pub fn code_span(source: &str) -> Option<TokenResult> {
  lazy_static! {
    static ref RULE: Regex = Regex::new(rule::CODE_SPAN_RULE).unwrap();
  }
  if let Some(caps) = RULE.captures(source) {
    let starting = caps.get(1).unwrap().as_str();
    let code = caps.get(2).unwrap().as_str();
    let ending = caps.get(3).unwrap().as_str();
    if code.is_empty() || starting.len() != ending.len() {
      return None;
    }
    let size = caps.get(0).unwrap().as_str().len();
    Some(TokenResult {
      size,
      token: Token::CodeSpan(code),
    })
  } else {
    None
  }
}

#[test]
fn test_code_span() {
  let cases = vec!["`12`", "``", "` `", "``123``", "``123```"];
  let mut results = vec![];
  for case in cases {
    results.push(code_span(case));
  }
  insta::assert_yaml_snapshot!(results);
}
