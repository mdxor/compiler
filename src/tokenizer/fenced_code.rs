use crate::tokenizer::rule;
use crate::tokenizer::token::*;
use lazy_static::lazy_static;
use regex::Regex;

fn split_once(target: &str) -> (&str, &str) {
  let mut slice1 = target;
  let mut slice2 = "";
  if let Some(index) = target.find(" ") {
    slice1 = &target[..index];
    slice2 = &target[index + 1..]
  }
  (slice1, slice2)
}

pub fn fenced_code(source: &str) -> Option<TokenResult> {
  lazy_static! {
    static ref HEAD_RULE: Regex = Regex::new(rule::FENCED_CODE_HEAD_RULE).unwrap();
    static ref NORMAL_RULE: Regex = Regex::new(rule::FENCED_CODE_BODY_RULE).unwrap();
  }
  if let Some(caps) = HEAD_RULE.captures(source) {
    let head = caps.get(0).unwrap().as_str();
    let head_size = head.len();
    let backtick_size = caps.get(1).map_or(0, |v| v.as_str().len());
    let tilde_size = caps.get(3).map_or(0, |v| v.as_str().len());
    let mut lang_meta = "";
    if backtick_size > 0 {
      lang_meta = caps.get(2).unwrap().as_str();
    } else {
      lang_meta = caps.get(4).unwrap().as_str();
    }
    lang_meta = lang_meta.trim();
    let (language, meta_string) = split_once(lang_meta);

    let caps_option = if backtick_size == 3 {
      NORMAL_RULE.captures(&source[head_size..])
    } else {
      let mut regex_string = String::from(r"(?:|([\s\S]*?)\n) {0,3}");
      if backtick_size > 0 {
        regex_string.push_str("`{");
        regex_string.push_str(&backtick_size.to_string());
      } else {
        regex_string.push_str(&"~");
        regex_string.push_str(&tilde_size.to_string());
      }
      regex_string.push_str(",} *(?:\n|$)|$");
      let regex = Regex::new(&regex_string).unwrap();
      regex.captures(&source[head_size..])
    };
    if let Some(caps) = caps_option {
      let size = caps.get(0).unwrap().as_str().len();
      let code = caps.get(1).map_or("", |v| v.as_str());
      Some(TokenResult {
        size: head_size + size,
        token: Token::FencedCode(FencedCode {
          code,
          language,
          meta_string,
        }),
      })
    } else {
      Some(TokenResult {
        size: source.len(),
        token: Token::FencedCode(FencedCode {
          code: &source[head_size..],
          language,
          meta_string,
        }),
      })
    }
  } else {
    None
  }
}

#[test]
fn test_fenced_code() {
  let cases = vec!["```ruby meta\ndef foo(x)\n  return 3\nend\n```"];
  let mut results = vec![];
  for case in cases {
    results.push(fenced_code(case));
  }
  insta::assert_yaml_snapshot!(results);
}
