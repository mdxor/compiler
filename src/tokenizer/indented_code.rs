use crate::tokenizer::rule;
use crate::tokenizer::token::*;
use lazy_static::lazy_static;
use regex::Regex;
pub fn indented_code(source: &str) -> Option<TokenResult> {
  lazy_static! {
    static ref RULE: Regex = Regex::new(rule::INDENTED_CODE_RULE).unwrap();
  }
  if let Some(caps) = RULE.captures(source) {
    let raw = caps.get(0).unwrap().as_str();
    let size = raw.len();
    let mut codes = vec![];
    let lines = raw.lines();
    for line in lines {
      codes.push(&line[4..]);
    }
    Some(TokenResult {
      size,
      token: Token::IndentedCode(IndentedCode { codes }),
    })
  } else {
    None
  }
}

#[test]
fn test_indented_code() {
  let cases = vec![
    "    a simple\n      indented code block",
    "    chunk1\n      \n      chunk2",
  ];
  let mut results = vec![];
  for case in cases {
    results.push(indented_code(case));
  }
  insta::assert_yaml_snapshot!(results);
}
