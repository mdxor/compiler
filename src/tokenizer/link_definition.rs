use crate::tokenizer::rule;
use crate::tokenizer::token::*;
use lazy_static::lazy_static;
use regex::Regex;
pub fn link_definition(source: &str) -> Option<TokenResult> {
  lazy_static! {
    static ref RULE: Regex = Regex::new(
      &rule::LINK_DEFINITION_RULE
        .replace("label", rule::LABEL_RULE)
        .replace("title", rule::TITLE_RULE)
    )
    .unwrap();
  }
  if let Some(caps) = RULE.captures(source) {
    let label = caps.get(1).unwrap().as_str().trim();
    if label == "" {
      return None;
    }
    let raw = caps.get(0).unwrap().as_str();
    let size = raw.len();

    let href = caps.get(2).unwrap().as_str();
    let mut title = caps.get(3).map_or("", |v| v.as_str());
    if !title.is_empty() {
      title = &title[1..title.len() - 1];
    }
    Some(TokenResult {
      size,
      token: Token::LinkDefinition(LinkDefinition { label, href, title }),
    })
  } else {
    None
  }
}

#[test]
fn test_link_definition() {
  let cases = vec![r#"[foo]: /url "title""#];
  let mut results = vec![];
  for case in cases {
    results.push(link_definition(case));
  }
  insta::assert_yaml_snapshot!(results);
}
