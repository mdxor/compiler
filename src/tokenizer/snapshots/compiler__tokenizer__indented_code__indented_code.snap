---
source: src/tokenizer/indented_code.rs
expression: results

---
- size: 38
  token:
    IndentedCode:
      codes:
        - a simple
        - "  indented code block"
- size: 30
  token:
    IndentedCode:
      codes:
        - chunk1
        - "  "
        - "  chunk2"

